### ISSUE TITLE
bug in animator dispose

### ISSUE THREAD

**Issue by sskodje** [XamlAnimatedGif/XamlAnimatedGif#21]
hey again!

This one got me on a bug hunt for over an hour :)

In Animator class, there is a couple small errors that in some situations causes a crash when disposing.

First off, in Animator.Play():

```
        public async void Play()
        {
 //this line below gets called on every play, even when _isStarted is true.
// This means that if you play an animator twice, when cancel() is called on the token in dispose,
// it gets called on a the new and unused token source.
// it needs to be moved inside the "if (!_isStarted)" statement.
            _cancellationTokenSource = new CancellationTokenSource();
            try
            {
                if (!_isStarted)
                {
                    _isStarted = true;
                    if (_timingManager.IsPaused)
                        _timingManager.Resume();
                    await RunAsync(_cancellationTokenSource.Token);
                }
                else if (_timingManager.IsPaused)
                {
                    _timingManager.Resume();
                }
            }
            catch (OperationCanceledException)
            {
            }
            catch (Exception ex)
            {
                AnimationBehavior.OnError(_image, ex, AnimationErrorKind.Rendering);
            }
        }
```

Then, in dispose:
```
        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                if (disposing)
                {
                    _cancellationTokenSource?.Cancel();
// I think that here, some times dispose is called on _sourceStream before the Task 
// has thrown on the cancellationtoken. This leads to exceptions when the task 
// tries to read from the stream.
//Probably a small delay of some milliseconds is enough.
                    _sourceStream.Dispose();
                }
                _disposed = true;
            }
        }
```

**Comment by thomaslevesque** [comment]
Hi sskodje,

Thanks for investigating the issue!

You're right, the `CancellationTokenSource` initialization is clearly incorrect, I'll fix it. Regarding the issue in `Dispose`, the problem probably occurs when we dispose the stream before the code using it had a chance to take the cancellation into account. I'm not comfortable adding a delay there, because we don't know how long it will take for the rendering thread to react to the cancellation, so it feels like a hack. Another, cleaner option it to wait for the task to complete. That's probably what I'll do.

btw, could you please post the exception details of when it crashes? I suspect the crash is due not to the exception itself, but to the fact that `_image` was not properly initialized, which caused a `NullReferenceException` when it tried to raise the `Error` event. I fixed that yesterday at the same time as the other bug you reported. Are you using the new release?

**Comment by sskodje** [comment]
Yeah, you're right. It was me getting stuck on a first chance exception, then a null reference exception. You new release works.

**Comment by thomaslevesque** [comment]
Fixed in 1.0.2
